modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -0,0 +1,337 @@
+      // store two nodes of each side
+      boundary_node_assm.push_back(std::make_pair(mesh.elem_ptr(std::get<0>(side_list_tmp[i]))
+                                                      ->side_ptr(std::get<1>(side_list_tmp[i]))
+                                                      ->node_id(0),
+                                                  mesh.elem_ptr(std::get<0>(side_list_tmp[i]))
+                                                      ->side_ptr(std::get<1>(side_list_tmp[i]))
+                                                      ->node_id(1)));
+    }
+  }
+  // Start from the first element, try to find a chain of nodes
+  boundary_orderred_node_list.push_back(boundary_node_assm.front().first);
+  boundary_orderred_node_list.push_back(boundary_node_assm.front().second);
+  // Remove the element that has been added to boundary_orderred_node_list
+  boundary_node_assm.erase(boundary_node_assm.begin());
+  const unsigned int boundary_node_assm_size_0 = boundary_node_assm.size();
+  for (unsigned int i = 0; i < boundary_node_assm_size_0; i++)
+  {
+    // Find nodes to expand the chain
+    dof_id_type end_node_id = boundary_orderred_node_list.back();
+    auto isMatch1 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair) {
+      return old_id_pair.first == end_node_id;
+    };
+    auto isMatch2 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair) {
+      return old_id_pair.second == end_node_id;
+    };
+    auto result = std::find_if(boundary_node_assm.begin(), boundary_node_assm.end(), isMatch1);
+    bool match_first;
+    if (result == boundary_node_assm.end())
+    {
+      match_first = false;
+      result = std::find_if(boundary_node_assm.begin(), boundary_node_assm.end(), isMatch2);
+    }
+    else
+    {
+      match_first = true;
+    }
+    // If found, add the node to boundary_orderred_node_list
+    if (result != boundary_node_assm.end())
+    {
+      boundary_orderred_node_list.push_back(match_first ? (*result).second : (*result).first);
+      boundary_node_assm.erase(result);
+    }
+    // If there are still elements in boundary_node_assm and result ==
+    // boundary_node_assm.end(), this means the boundary is not a loop, the
+    // boundary_orderred_node_list is flipped and try the other direction that has not
+    // been examined yet.
+    else
+    {
+      if (isFlipped)
+      {
+        // Flipped twice; this means the boundary has at least two segments.
+        return false;
+      }
+      // mark the first flip event.
+      isFlipped = true;
+      std::reverse(boundary_orderred_node_list.begin(), boundary_orderred_node_list.end());
+      // As this iteration is wasted, set the iterator backward
+      i--;
+    }
+  }
+  // If the code ever gets here, boundary_node_assm is empty.
+  // If the isFlipped == true, the boundary is not a loop.
+  // This is not done inside the loop just for some potential applications in the future.
+  if (isFlipped)
+    return false;

```

Comment:
why is it not a loop if it's flipped here? What's the point of retrying the with the reverse boundary nodes then?

Reply:
A closed loop should have no ends so that flipping is not needed to traverse all the nodes. For an open curve, if we do not start from one end, then we will need to "flip" once to traverse all the nodes. I try to traverse all the nodes after flipping for some possible future applications.